processing/src/main/java/io/druid/query/extraction/SubstringDimExtractionFn.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+import javax.annotation.Nullable;
+
+/**
+ */
+public class SubstringDimExtractionFn extends DimExtractionFn
+{
+  private static final byte CACHE_TYPE_ID = 0x8;
+
+  private final int index;
+  private final int length;
+  private final int end;
+
+  @JsonCreator
+  public SubstringDimExtractionFn(
+      @JsonProperty("index") int index,
+      @Nullable
+      @JsonProperty("length") int length
```

Comment:
@Nullable won't do anything here, null will be mapped to 0. Instead we should make thing argument `Integer` object and set length to `-1` if it's null so we can distinguish length = 0 from null.